lib/collection/src/shards/forward_proxy_shard.rs

```diff
@@ -437,8 +478,44 @@ impl ShardOperation for ForwardProxyShard {
             };
 
             op.map(|op| OperationWithClockTag::new(op, tag))
+        } else if let Some(filter) = &self.filter {
```

Comment:
I vote to move both branches into a separate function, since this function is already 150+ lines.

Using an enum to either use resharding or a filter here would also be helpful here.